lib/ansible/modules/cloud/vmware/vmware_maintenancemode.py

```diff
@@ -97,109 +101,96 @@
-
-    try:
-        task = host.EnterMaintenanceMode_Task(
-            module.params['timeout'],
-            module.params['evacuate'],
-            spec)
-
-        success, result = wait_for_task(task)
-
-        return dict(changed=success,
-                    hostsystem=str(host),
-                    hostname=module.params['esxi_hostname'],
-                    status='ENTER',
-                    msg='Host entered maintenance mode')
-
-    except TaskError:
-        module.fail_json(
-            msg='Host failed to enter maintenance mode')
-
-
-def ExitMaintenanceMode(module, host):
-    if not host.runtime.inMaintenanceMode:
-        module.exit_json(
-            changed=False,
-            hostsystem=str(host),
-            hostname=module.params['esxi_hostname'],
-            status='NO_ACTION',
-            msg='Host not in maintenance mode')
-
-    try:
-        task = host.ExitMaintenanceMode_Task(
-            module.params['timeout'])
-
-        success, result = wait_for_task(task)
-
-        return dict(changed=success,
-                    hostsystem=str(host),
-                    hostname=module.params['esxi_hostname'],
-                    status='EXIT',
-                    msg='Host exited maintenance mode')
-
-    except TaskError:
-        module.fail_json(
-            msg='Host failed to exit maintenance mode')
+from ansible.module_utils.vmware import PyVmomi, TaskError, vmware_argument_spec, wait_for_task
+from ansible.module_utils._text import to_native
+
+
+class VmwareMaintenanceMgr(PyVmomi):
+    def __init__(self, module):
+        super(VmwareMaintenanceMgr, self).__init__(module)
+        self.esxi_hostname = self.module.params.get('esxi_hostname')
+        self.vsan = self.module.params.get('vsan', None)
+        self.host = self.find_hostsystem_by_name(host_name=self.esxi_hostname)
+        if not self.host:
+            self.module.fail_json(msg='Host %s not found in vCenter' % self.esxi_hostname)
+
+    def EnterMaintenanceMode(self):
+        if self.host.runtime.inMaintenanceMode:
+            self.module.exit_json(changed=False,
+                                  hostsystem=str(self.host),
+                                  hostname=self.esxi_hostname,
+                                  status='NO_ACTION',
+                                  msg='Host %s already in maintenance mode' % self.esxi_hostname)
+
+        spec = vim.host.MaintenanceSpec()
+
+        if self.vsan:
+            spec.vsanMode = vim.vsan.host.DecommissionMode()
+            spec.vsanMode.objectAction = self.vsan
+
+        try:
+            task = self.host.EnterMaintenanceMode_Task(self.module.params['timeout'],
+                                                       self.module.params['evacuate'],
+                                                       spec)
+
+            success, result = wait_for_task(task)
+
+            return self.module.exit_json(changed=success,
```

Comment:
I think the `return` isn't necessary (you don't use it in ExitMaintenanceMode).

Reply:
My bad, missed it.